docs/osf-trolling-list/build_process.md

```diff
@@ -0,0 +1,127 @@
+
+This is because knowing hashes of firmware components we can calculate expected
+TPM PCRs values. Knowing good PCR values and being able to predict further
+values after firmware update, we gaining following advantages:
+
+* We can easily confirm if firmware is valid by reading TPM PCRs and compare to
+  reference integrity measures (RIM), RIM can be calculated by build exactly
+  the same version of OSF
+* Process of replying measurements provided in TPM Event Log is simpler because
+  we can easily obtain software and confirm hashes used during PCR extension
+  process were valid and the same as produced by software build process
+* Looking for bugs by using bisection is easier since being able to build every
+  commit of software in reproducible manner give us confidence that there would
+  be no change in behavior between 2 consecutive builds
+* Consistent behavior for given hash can be confirmed by test results tied to
+  given hash in that way users looking at test results may expect exactly the
+  same behavior for firmware binaries with the same hash
+* Long term maintenance should be easier if firmware build would be
+  reproducible over long time, because often happen some firmware land in
+  industrial applications e.g. robots, trains, smart city infrastructure, if
+  bug will be detected after long time it is important to have tools and be
+  able to confirm the same binary can be generated in future. If we know given
+  toolchain reproduce binary bit by bit we are sure it also reproduce
+  software/firmware behavior. In that light making small incremental
+  improvement using toolchain that give reproducible results give us higher
+  chance that we will not introduce uncontrolled change in behavior and code
+  change will be reflected by compilation process without affecting previous
+  software behavior.
+
+# [OSFI0011] Docker containers as build systems "considered harmful"
+
+Docker containers for couple last years become default method for provide
+reproducible runtime environment for software as well those started to be very
+useful for developers to transfer exact configuration for building given
+software stack. We wrote about usefulness of Docker containers in embedded
+software development environment on 
+[3mdeb blog](https://blog.3mdeb.com/2018/2018-09-27-optimize-performance-in-docker/).
+
+After long time of using Docker containers for embedded software development
+and build environments we noticed problems for long time maintenance and
+reproducibility.
+
+## coreboot-sdk problems
+
+Example can be `coreboot-sdk` used for building [coreboot project](https://coreboot.org).
+
+Good summary of the problem was provided by [Thrilleratplay in guix-docker repo](https://github.com/Thrilleratplay/guix-docker#the-problem):
+
+> Currently, the coreboot build environment, `coreboot-sdk`, uses a Debian docker
+> base image.  To install additional required packages, `apt-get update` must be
```

Comment:
This problem could have been resolved by providing a custom Docker image where the required packages are already pre-installed and everything is ready for a coreboot build.

Reply:
> we have to keep the image forever because it cannot be rebuilt

It could be rebuilt, as long as these older pre-compiled .deb packages used for the image _(and/or their sources)_ - are still available. Actually, if we insist on a "live rebuild" of an image _(i.e. to save space)_, "apt-get install X" could be replaced with "wget https://link_to_archieved/older_X.deb && dpkg -i ./older_X.deb". Although there could be occasional dependency "older X depends on older Y" problems depending on how an upstream basic Docker image develops, in theory this is possible.

Reply:
@mikebdp2 I added both ideas as potential problem solutions. In the future feel free to suggest improvements in the text - it would be easier to move things forward.